### DISCUSSION TITLE
Change in semantics and policy for Perl updating

### DISCUSSION THREAD

**Discussion by ehuelsmann** [renovatebot/renovate#23762] [Request Help] [answered]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

Placeholder value, please select the correct response from the dropdown

### Describe the problem

Let me start by saying that I'm very, very happy user of Mend Renovate for *years* now, with respect to dependencies managed in the JavaScript world.

Some time ago, Renovate started to support Perl dependency identification *and* enabled it by default (opt-out). Although this might seem like a good idea, this is actually a problem, because the Perl world is very, *very* different from the JavaScript world. Where I may see weekly releases of JavaScript projects, with minor and even major version updates on a weekly basis, the Perl world works differently. Many Perl module authors take a lot of care for their code to keep running on version v5.8.0 (released 2002). The same is true for modules being released: authors take a *lot* of care to specify the lowest dependency version which their software will run on. Often, this means that the version specified is one or more *decades* old.

How does this relate back to Renovate? Well, Perl modules specify the oldest version of Perl they will support as one of their dependencies; you might see the following in a `cpanfile`:

```perl
requires 'perl' => '5.14.0';
```
In addition, I should note that Perl version numbers are weird things. These are all the same:
* 5.14.0
* 5.014
* v5.14 (I *think*)

But `5.14` is not the same: it's actually `5.140`. There's more on Perl version number parsing and comparing here: https://metacpan.org/dist/version/view/lib/version.pm#PARSING-AND-COMPARING-VERSIONS

The version numbers specify the intention of the author that users with older Perl versions should be able to run this module. Changing this number to the latest released Perl (we see updates to e.g. `5.39.1`) means that people suddenly can't install these modules on their OS anymore without installing e.g. Perl 5.39.1. However, replacing a system Perl is a tedious operation that's not for the faint-of-heart; in many cases Perl is tied to system software and upgrading should only be done by the developers of the distribution.
There's another problem with the updates - this relates specifically to the Perl version numbers you're using - to the Perl version numbers: you seem to be using odd numbered minor versions; e.g. 5.37.0 and 5.39.1. These are development snapshots and *should absolutely be excluded* from version number specifications.

In summary:
* Perl ecosystem is very different from JavaScript/Golang/...
* Perl versions indicate minimum required versions, which do not change when new versions are released
* Perl module versions indicate the same

I'd love to go into further discussion about this, but for now I have to conclude that Perl version updates should be opt-in instead of opt-out: It's *great* that renovate supports it, but for most projects it isn't a good default to update Perl dependency versions.

### Have you created a minimal reproduction repository?

I have explained in the description why a minimal reproduction is impossible

**Comment by mpkorstanje** [comment]
> I'd love to go into further discussion about this, but for now I have to conclude that Perl version updates should be opt-in instead of opt-out: It's great that renovate supports it, but for most projects it isn't a good default to update Perl dependency versions.

@ehuelsmann from the perspective of Renovate, this behavior is opt in. And in the Cucumber org we've opted in by using `:automergePatch` and `:automergeMinor` presets. So it is reasonable to say that we've made this choice.

However the Cucumber org is probably unique in that a repository typically contains multiple different languages. So for us it is important that the semantics of JavaScript align with those of Perl. So I think a more suitable request would be to recognize that semantically for the perl dependency, minor versions are actually major versions and patch versions are actually minor versions (though as you've mentioned, it's actually more complicated than that with the odd-even structure).

**Reply by ehuelsmann** [reply]
Well, the Cucumber org isn't really unique in this respect: although I think that the way Cucumber combines languages is indeed unique, it's far less unique to have a combination of any language in combination with JavaScript, because that's where half the world seems to have moved its efforts.

The LedgerSMB project is another project that I'm involved in; it's a combination of Perl and JavaScript. For the JavaScript we want the dependency management offered by Renovate *very much*. However, for the Perl side we took great care to identify - not just for Perl, but for all the dependencies, just like most Perl projects - which modules are minimum versions we depend on, taking APIs and bugfixes into account. In that project, I immediately disabled "cpanfile" as a source for dependencies. Don't get me wrong. I think having the "cpanfile" source is tremendously useful for compliance with corporate policies (honest remark). It's just not how most *open source* Perl projects are being run.

I understand what you say about it being opt-in from the Renovate perspective. If that's where you are coming from, then I mean that - for Perl open source projects - this really should be double opt-in: enabling renovate and on top of that explicitly adding `cpanfile` as a source for dependencies, because it's not a good fit.

I'd like to stress again that I'm absolutely fan and happy user of Renovate and that I'm totally just trying to find the most helpful way forward, because, actually, if it not had been for the fact that "cpanfile" is enabled by default, I wouldn't have known Renovate supports it.

**Comment by rarkins** [comment]
@ikesyo you wrote the `cpanfile` manager - what do you think about setting a skipReason for `perl` so that it's not updated?

Regarding some of the above points:

Perl versioning and it's confusing nature seems to be an independent topic to whether perl itself should be updated by default. Let's instead stick to whether perl should be updated or not.

I don't want to introduce any concept like "minor versions are actually major versions" in Renovate. Instead, the concept of breaking/not breaking should be introduced which can be independent of major/minor/patch. It will simply be confusing to tell people "minor are major", when in fact what people care about is the breaking nature.

Once a dependency is skipped by default, we don't currently offer the ability to unskip it, although we do want to introduce that.

**Reply by ehuelsmann** [reply]
Right. I don't think it's a good idea to adopt the idea that "minor is major": although Perl 7 is out of sight, it *might* still happen some day. Just not in the foreseeable future. If that would be the case, then what's the definition of minor versus major? I don't know how to explain that to software at that point...

The thing that I tried to explain in my response above to mpkorstanje is that this idea of specifying minimum versions -- not latest versions -- does not just apply to Perl itself, but also to all the other modules that Perl projects specify in their cpanfile.

Well, at least, that's my experience. I'm kind of hoping that other voices jump in either to confirm what I'm saying or to declare that their situation is totally different.

**Comment by mpkorstanje** [comment]
>  Instead, the concept of breaking/not breaking should be introduced which can be independent of major/minor/patch. It will simply be confusing to tell people "minor are major", when in fact what people care about is the breaking nature.

Seems fair. 

I've currently disabled the perl dependency in the cpanfile manager in our own config.

```
  "packageRules": [
    {
      "matchManagers": [ "cpanfile"],
      "matchPackageNames": ["perl"],
      "enabled": false
    }
  ]
```

Perhaps this could be made a default or a preset?

**Reply by ehuelsmann** [reply]
This is absolutely a good preset (to begin with).

**Reply by viceice** [reply]
this would allow users how build applications to opt in to perl updates if they don't need to care about other users.

**Reply by ehuelsmann** [reply]
Exactly. I'd still say that odd-numbered minors need to be excluded in the updates as they are development snapshots, though.

**Reply by viceice** [reply]
this sounds like we need a perl versioning which treats those versions as unstable, like we do in Ubuntu versioning for non lts versions

**Reply by HonkingGoose** [reply]
Do the maintainers want a new `workarounds:disablePerlUpdates` preset then?

```json
{
  "packageRules": [
    {
      "matchManagers": ["cpanfile"],
      "matchPackageNames": ["perl"],
      "enabled": false
    }
  ]
}
```

The new preset should _not_ be put in the `extends` array of the `workarounds:all` preset, because then we're disabling Perl for all users.

**Reply by rarkins** [reply]
Yes, that would be good

**Reply by HonkingGoose** [reply, marked as answer]
See issue:

- https://github.com/renovatebot/renovate/issues/23900
